File: SoulLink_v1.5.4/backend/app/models/relationship.py
# /backend/app/models/relationship.py
# /version.py
# /_dev/

# "War, war never changes."
# - Fallout
from sqlmodel import SQLModel, Field, Column
from pydantic import BaseModel
from sqlalchemy import JSON, UniqueConstraint
from datetime import datetime
from typing import Optional, Dict, Any

# "Had to be me. Someone else might have gotten it wrong."
# - Mordin Solus - Mass Effect
class SoulRelationship(SQLModel, table=True):
    __table_args__ = (UniqueConstraint("user_id", "soul_id", name="uq_user_soul"),)
    """
    Tracks the bond between a specific User and a Soul.
    Handles intimacy progression, location state, and special flags.
    """
    __tablename__ = "user_soul_relationships"

    relationship_id: Optional[int] = Field(default=None, primary_key=True)

    # Core linkage
    # Ensure foreign_key="users.user_id" matches your actual User model table name
    user_id: str = Field(
        index=True,
        max_length=36 
    )
    soul_id: str = Field(
        index=True,
        max_length=50
    )

    # Intimacy system
    intimacy_score: int = Field(default=0)
    intimacy_tier: str = Field(
        default="STRANGER",
        max_length=20,
        description="STRANGER | TRUSTED | SOUL_LINKED"
    )

    # Current world state
    current_location: Optional[str] = Field(
        default="linkgate_mall", # Default starting point
        max_length=50
    )

    # Soul-specific flags (curse progress, glove status, etc.)
    # We use Dict[str, Any] to match your Soul model pillar style
    special_flags: Dict[str, Any] = Field(
        default_factory=dict,
        sa_column=Column(JSON)
    )

    # 🚀 RECOGNITION & CONTENT GATING 🚀
    # Set to True only if user matches the dev_config during first link or via admin
    is_architect: bool = Field(default=False)
    
    # Decides if NSFW content is active for this specific pairing
    nsfw_unlocked: bool = Field(default=False)

    # "Don't be sorry, be better."
    # - Kratos - God of War
    created_at: datetime = Field(default_factory=datetime.utcnow)
    last_interaction: datetime = Field(default_factory=datetime.utcnow, sa_column_kwargs={"onupdate": datetime.utcnow})

# --- READ SCHEMAS (For API Responses) ---
class RelationshipRead(BaseModel):
    soul_id: str
    intimacy_tier: str
    current_location: str
    is_architect: bool
    nsfw_unlocked: bool
    last_interaction: datetime

File: SoulLink/SoulLink_v1.5.3/frontend/lib/models/relationship.dart
// frontend/lib/models/relationship.dart
// version.py v1.5.3-P

class SoulRelationship {
  final String soulId;
  final String name;
  final String archetype;
  final String currentLocation;
  final String intimacyTier;
  final int intimacyScore;
  final bool isArchitect;
  final bool nsfwUnlocked;
  final String lastInteraction;
  final String portrait_url; // 🖼️ THE MISSING LINK

  SoulRelationship({
    required this.soulId,
    required this.name,
    required this.archetype,
    required this.currentLocation,
    required this.intimacyTier,
    required this.intimacyScore,
    required this.isArchitect,
    required this.nsfwUnlocked,
    required this.lastInteraction,
    required this.portrait_url, // Added to constructor
  });

  factory SoulRelationship.fromJson(Map<String, dynamic> json) {
    return SoulRelationship(
      soulId: json['soul_id'] ?? "",
      name: json['name'] ?? "Unknown",
      archetype: json['archetype'] ?? "Unknown",
      currentLocation: json['location'] ?? "Unknown",
      intimacyTier: json['tier'] ?? "STRANGER",
      intimacyScore: json['intimacy_score'] ?? 0,
      isArchitect: json['is_architect'] ?? false,
      nsfwUnlocked: json['nsfw_unlocked'] ?? false,
      lastInteraction: json['last_interaction'] ?? "",
      // ✅ Map the backend field to the model
      portrait_url: json['portrait_url'] ?? "/assets/images/souls/default.jpg",
    );
  }
}